text/0000-unused_const_fn_result_lint.md

```diff
@@ -0,0 +1,173 @@
+checks for unused results of const fn invocations.
+
+The end result is the same, but now the compiler verifies automatically
+whether the function is actually side effect free,
+and no manual annotations on the functions are needed beyond `const fn`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler should get a new lint called `unused_const_fn_result` which
+is meant to fire if you invoke a `const fn` function but discard the result.
+
+If you write code like:
+
+```rust
+const fn add_one(v: u32) -> u32 {
+    v + 1
+}
+fn foo() {
+    add_one(2);
+}
+```
+The lint will fire, pointing to the invocation of `add_one`.
+It shall suggest you to remove the entire function call.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `unused_const_fn_result` lint shall only consider function calls of functions
+applicable for linting that:
+
+* are `const fn`
+* don't have uninhabited types as a return type (`enum Void {}`, `!`)
+
+The first criterion gives the lint its name, and contributes a large
+component in ensuring that the call actually is side effect free,
+but the other criteria are important as well.
+
+The second criterion is needed as divergence is a clear side effect.
+Also, as the never type `!` affects the type of the enclosing function,
+removing the invocation might might stop the code from compiling.
+
+Of course, one might argue that functions that never return despite
+having a return type distinct from an uninhabited one, e.g. by
+running an infinite computation but still claiming to return `u32`,
+qualify as side effect causing as well. However, finding out
+whether a call terminates or not is precisely the halting problem
+which is undecidable as famously shown.
+From a practical standpoint, it is of low relevance whether
+a function call takes 3 googol years to evaluate, or infinitely
+long. Optimizers already expect that they are allowed to remove
+dead code with finite nonzero execution time (and in C,
+infinite as well), so relying on this for side effects
+is quite brittle.
+
+Additionally, in order for the lint to fire on a concrete function call, the call must:
+
+* have all its generic types known... no generic types might be contributed from the calling context
+* have the types of all input parameters contain no mutability (`&mut` references, `* mut` pointers, `UnsafeCell`)
+* have the types of all input params which are moved have a side effect free `Drop` impl
+
+The first two criteria are to prevent the lint from firing for function invocations which still have mutable invocations.
+A simple example (inspired by [@rkruppe's example](https://github.com/rust-lang/rust/pull/50805#issuecomment-389654872)):
+
+
+```rust
+trait AddOne {
+    fn add_one(self);
+}
+
+impl<'a> AddOne for &'a mut u32 {
+    fn add_one(self) {
+        *self += 1;
+    }
+}
+
+const fn add_one<T: const AddOne>(v: T) {
+    v.add_one();
+}
+```
```

Comment:
Which of the two criteria does this example fall under? Remember `()` is inhabited.

Reply:
"have all its generic types known" and "have the types of all input parameters contain no mutability". I re-started the numbering for the second group of criteria, but this seems to be confusing, so I might change that.